### PR TITLE
Fix issue of zero intensity from IntegrateSphere when using ellipsoid peaks

### LINKED ISSUE #30898: IntegratePeaksMD return zero intensity for some peaks when using ellipsoidal integration
When using new ellipsoidal peak integration capability in `IntegratePeaksMD`, some peak intensities are returned as zero. When using the default spherical integration, the same behavior is not observed.

Original issue tracked on Gitlab: [#226](https://code.ornl.gov/sns-hfir-scse/diffraction/single-crystal/single-crystal-diffraction/-/issues/226)

### PR DESCRIPTION
**Pull request by KedoKudo**
**Description of work.**

This PR implemented a new logic (inspired by `IntegartePeaksMD2.findEllipsoid`) for checking if child grid box need to be searched during integrating spheres/ellipsoids.

- Related Gitlab task: [#233](https://code.ornl.gov/sns-hfir-scse/diffraction/single-crystal/single-crystal-diffraction/-/issues/233) and [#249](https://code.ornl.gov/sns-hfir-scse/diffraction/single-crystal/single-crystal-diffraction/-/issues/249)
- This is #30897 into `master`

**To test:**

Use the following code the test

```python
from mantid.simpleapi import *
import numpy as np
ws = CreateMDWorkspace(Dimensions='3', Extents='0.5,1.5,0.5,1.5,0.5,1.5',
                       Names='Q_lab_x,Q_lab_y,Q_lab_z', Units='U,U,U',
                       Frames='QLab,QLab,QLab',
                       SplitInto='2', SplitThreshold='50')
expt_info = CreateSampleWorkspace()
ws.addExperimentInfo(expt_info)
nevents=10000
FakeMDEventData(ws, UniformParams=f'{nevents},0.98,1.02,0.98,1.02,0.98,1.02')
# add peak
p = CreatePeaksWorkspace(InstrumentWorkspace='ws', NumberOfPeaks=0, OutputWorkspace='peaks')
SetUB('peaks', 1,1,1,90,90,90)
pk = p.createPeak([1,1,1]) # axes aligned with viewing axes
p.addPeak(pk)
a = 0.005
b = 0.013
c = 0.013
#a=b=c=0.005
p=IntegratePeaksMD(InputWorkspace='ws', PeakRadius=[a,b,c], Ellipsoid=True, PeaksWorkspace='peaks', OutputWorkspace='ellipsoid')
print("Integrate Intensity =", p.getPeak(0).getIntensity())
print("Approx. Expected Intensity=", 4/3*np.pi*(a*b*c)*nevents*25**3)
```

Fixes #30898 . 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by stonecoldhughes — DISMISSED** [review]
This looks good. Well done!

File: Framework/DataObjects/inc/MantidDataObjects/MDGridBox.tcc
Comment (RichardWaiteSTFC) [inline Framework/DataObjects/inc/MantidDataObjects/MDGridBox.tcc:1285]: Could we not do the partial box check but just using the major axis length for the spherical radius? Would that give a speedup?

Comment (KedoKudo) [reply Framework/DataObjects/inc/MantidDataObjects/MDGridBox.tcc:1285]: That's essentially what we are doing here. Given that this is a temporary solution (until we find a better way to check the partial box for ellipsoid) and checking partial box doesn't take a lot of time, this is the minimum amount of changes we can make to ensure it can work for both spherical peaks and ellipsoid peaks.

**Comment by rosswhitfield** [comment]
This fix appears to be really slow. I see IntegratePeaksMD going from less than 0.01 seconds to 1.5 seconds with this fix.

* Can you change the workflow to only apply the fix for the case when you manually specify the three radii and not when running `findEllipsoid` so this won't affect ISIS.
* Can you add a unit test that captures the defect.
* Can you add something to the release notes [here](https://docs.mantidproject.org/nightly/release/v6.1.0/diffraction.html#known-defects) about this temporary slow fix.

**Comment by RichardWaiteSTFC** [comment]
> This fix appears to be really slow. I see IntegratePeaksMD going from less than 0.01 seconds to 1.5 seconds with this fix.
> 
> * Can you change the workflow to only apply the fix for the case when you manually specify the three radii and not when running `findEllipsoid` so this won't affect ISIS.
> * Can you add a unit test that captures the defect.
> * Can you add something to the release notes [here](https://docs.mantidproject.org/nightly/release/v6.1.0/diffraction.html#known-defects) about this temporary slow fix.

I don't know whether this helps but I recently added a check to see if a box was definitely NOT contained (so slightly different thing) inside a spherical region in IntegratePEaksMD to avoid the number of boxes I needed to loop over all events when determining the covariance matrix
https://github.com/mantidproject/mantid/pull/30800/files#diff-9f7b0ca46aab4d667a50278bc89941b5a4adf603d97ee5cd14a81334f3538b6aR1010
I know you don't see your problem in MDGridBox.tcc with spherical regions, but I believe this rough check with the major axis length squared of an ellipsoid would cut down the number of boxes we need to loop over (i.e. some boxes will be accepted that are not actually partially contained, but no boxes that are partially contained will be missed if that makes sense...)

**Comment by KedoKudo** [comment]
> This fix appears to be really slow. I see IntegratePeaksMD going from less than 0.01 seconds to 1.5 seconds with this fix.
> 
> * Can you change the workflow to only apply the fix for the case when you manually specify the three radii and not when running `findEllipsoid` so this won't affect ISIS.

The fix now only affects cases where
- Ellipsoid is explicitly specified
- Three radii are provided

> * Can you add a unit test that captures the defect.

@stonecoldhughes is working on updating the unit tests

> * Can you add something to the release notes [here](https://docs.mantidproject.org/nightly/release/v6.1.0/diffraction.html#known-defects) about this temporary slow fix.

I added one sentence in the specified section and updated the PR description accordingly.

**Comment by rosswhitfield** [comment]
I like this approach better but using the a,b,c values originally reported in the defect, it hasn't fixed the issue.
```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

ws = CreateMDWorkspace(Dimensions='3', Extents='0.5,1.5,0.5,1.5,0.5,1.5',
                       Names='Q_lab_x,Q_lab_y,Q_lab_z', Units='U,U,U',
                       Frames='QLab,QLab,QLab',
                       SplitInto='2', SplitThreshold='50')
expt_info = CreateSampleWorkspace()
ws.addExperimentInfo(expt_info)
nevents=100000000
FakeMDEventData(ws, UniformParams=nevents)
# add peak
p = CreatePeaksWorkspace(InstrumentWorkspace='ws', NumberOfPeaks=0, OutputWorkspace='peaks')
SetUB('peaks', 1,1,1,90,90,90)
pk = p.createPeak([1,1,1]) # axes aligned with viewing axes
p.addPeak(pk)
a = 0.005
b = 0.013
c = 0.013
p=IntegratePeaksMD(InputWorkspace='ws', PeakRadius=[a,b,c], Ellipsoid=True, PeaksWorkspace='peaks', OutputWorkspace='ellipsoid')
print("Integrated Intensity =", p.getPeak(0).getIntensity())
print("Approx. Expected Intensity=", 4/3*np.pi*(a*b*c)*nevents)
```
gives

```
Integrated Intensity = 0.0
Approx. Expected Intensity= 353.9527723044499
```

**Comment by KedoKudo** [comment]
> I like this approach better but using the a,b,c values originally reported in the defect, it hasn't fixed the issue.
> 
> ```
> # import mantid algorithms, numpy and matplotlib
> from mantid.simpleapi import *
> import matplotlib.pyplot as plt
> import numpy as np
> 
> ws = CreateMDWorkspace(Dimensions='3', Extents='0.5,1.5,0.5,1.5,0.5,1.5',
>                        Names='Q_lab_x,Q_lab_y,Q_lab_z', Units='U,U,U',
>                        Frames='QLab,QLab,QLab',
>                        SplitInto='2', SplitThreshold='50')
> expt_info = CreateSampleWorkspace()
> ws.addExperimentInfo(expt_info)
> nevents=100000000
> FakeMDEventData(ws, UniformParams=nevents)
> # add peak
> p = CreatePeaksWorkspace(InstrumentWorkspace='ws', NumberOfPeaks=0, OutputWorkspace='peaks')
> SetUB('peaks', 1,1,1,90,90,90)
> pk = p.createPeak([1,1,1]) # axes aligned with viewing axes
> p.addPeak(pk)
> a = 0.005
> b = 0.013
> c = 0.013
> p=IntegratePeaksMD(InputWorkspace='ws', PeakRadius=[a,b,c], Ellipsoid=True, PeaksWorkspace='peaks', OutputWorkspace='ellipsoid')
> print("Integrated Intensity =", p.getPeak(0).getIntensity())
> print("Approx. Expected Intensity=", 4/3*np.pi*(a*b*c)*nevents)
> ```
> 
> gives
> 
> ```
> Integrated Intensity = 0.0
> Approx. Expected Intensity= 353.9527723044499
> ```

The peak size above (0.005, 0.013, 0.013) falls below the smallest grid defined in `CreateMDWorkspace`, so the current version will simply skip over it and return 0.

**Comment by rosswhitfield** [comment]
But if I give `a=b=c=0.005` then I get the correct answer, which is even smaller.

**Comment by KedoKudo** [comment]
> But if I give `a=b=c=0.005` then I get the correct answer, which is even smaller.

Here is the function call I used
```
p=IntegratePeaksMD(InputWorkspace='ws', PeakRadius=0.005, PeaksWorkspace='peaks', OutputWorkspace='ellipsoid')
print(p.getPeak(0).getIntensity())
print(4/3*np.pi*(0.005**3)*nevents)
```
and I got 0 intensity as output, which is expected.

**Comment by rosswhitfield** [comment]
No. Did you actually run my script? I am guessing you are using the `nevents=10_000` from the issue description, which then it is right. Try `nevents=100000000`

**Comment by RichardWaiteSTFC** [comment]
I'm doing some comparisons on real (WISH) data and have noticed that when using findEllipsoid (i.e. not specifying all radii) - there are some peaks that can be integrated as spheres but the ellipsoid integration comes out as 0 - so it looks like there is a problem! FYI this test was done on v6 release.
image
On a separate note the ellipsoid intensities are slightly lower in general - this could be due to less background, or the peak is contaminating the bg shell - I need to keep investigating.

**Comment by rosswhitfield** [comment]
This appears to now be doing the correct thing. :joy:

This PR needs a bit of a cleanup. There is a merge conflict that needs to be resolved. I don't think you need `boost/optional.hpp` anymore. There are some random `.nfs*` files that should be removed.

File: Framework/MDAlgorithms/test/IntegratePeaksMD2Test.h
Comment (rosswhitfield) [inline Framework/MDAlgorithms/test/IntegratePeaksMD2Test.h:560]: We can make this test exactly reproduce the 0 intensity error:

```diff
--- a/Framework/MDAlgorithms/test/IntegratePeaksMD2Test.h
+++ b/Framework/MDAlgorithms/test/IntegratePeaksMD2Test.h
@@ -559,30 +559,45 @@ public:
 
   void test_exec_EllipsoidRadii_NoBackground_SingleCount_Vol_LongEllipse() {
 
+    CreateMDWorkspace algC;
+    TS_ASSERT_THROWS_NOTHING(algC.initialize())
+    TS_ASSERT(algC.isInitialized())
+    TS_ASSERT_THROWS_NOTHING(algC.setProperty("Dimensions", "3"));
+    TS_ASSERT_THROWS_NOTHING(
+        algC.setProperty("Extents", "-0.5,0.5,-0.5,0.5,-0.5,0.5"));
+    TS_ASSERT_THROWS_NOTHING(algC.setProperty("Names", "h,k,l"));
+    TS_ASSERT_THROWS_NOTHING(algC.setProperty("Units", "U,U,U"));
+    TS_ASSERT_THROWS_NOTHING(algC.setProperty("Frames", "HKL,HKL,HKL"));
+    TS_ASSERT_THROWS_NOTHING(algC.setProperty("SplitInto", "2"));
+    TS_ASSERT_THROWS_NOTHING(algC.setProperty("MaxRecursionDepth", "5"));
+    TS_ASSERT_THROWS_NOTHING(algC.setPropertyValue(
+        "OutputWorkspace", "IntegratePeaksMD2Test_MDEWS"));
+    TS_ASSERT_THROWS_NOTHING(algC.execute());
+    TS_ASSERT(algC.isExecuted());
+
     // Test an ellipsoid against theoretical vol
-    size_t numEvents = 2000000;
+    size_t numEvents = 200000;
     V3D pos(0.0, 0.0, 0.0); // peak position
 
-    createMDEW();
-
     // Major axis along x
     double fail_val = 0.013;
-    std::vector< double > radii = {0.05, fail_val, fail_val };
+    std::vector<double> radii = {0.005, fail_val, fail_val};
 
     Instrument_sptr inst =
         ComponentCreationHelper::createTestInstrumentCylindrical(5);
     PeaksWorkspace_sptr peakWS(new PeaksWorkspace());
-    addUniform(numEvents, {std::make_pair(-0.5, 0.5), std::make_pair(-0.5, 0.5),
-                           std::make_pair(-0.5, 0.5)});
+    addUniform(numEvents,
+               {std::make_pair(-0.05, 0.05), std::make_pair(-0.05, 0.05),
+                std::make_pair(-0.05, 0.05)});
     peakWS->addPeak(Peak(inst, 1, 1.0, pos));
     AnalysisDataService::Instance().addOrReplace("IntegratePeaksMD2Test_peaks",
                                                  peakWS);
 
-    double ellipVol = (4.0 / 3.0) * M_PI * static_cast<double>(numEvents) *
+    double ellipVol = (4.0 / 3.0) * M_PI *
+                      static_cast<double>(numEvents * 1000) *
                       std::accumulate(radii.begin(), radii.end(), 1.0,
                                       std::multiplies<double>());
 
     doRun(radii, {0.0}, "IntegratePeaksMD2Test_peaks_out", {0.0}, false, false,
           "NoFit", 0.0, true, false);
```

**Review by rosswhitfield — DISMISSED** [review]

File: Framework/DataObjects/inc/MantidDataObjects/MDGridBox.tcc
Comment (martyngigg) [inline Framework/DataObjects/inc/MantidDataObjects/MDGridBox.tcc:1167]: Just a quick question in terms of memory handling. Is there any reason we can't tidy up a bit here and use a vector rather than `new` arrays directly. Unless we can gurantee no exceptions are thrown until the `deletes` below then we could leak memory if something throws?

For example,
```cpp
auto verticesContained = std::vector<size_t>(numBoxes, 0); 
```

should do the same as 

```cpp
auto verticesContained = new size_t[numBoxes];
memset(verticesContained, 0, numBoxes * sizeof(size_t));
```
and be safer? And then the same for all of the other bare arrays?

**Review by martyngigg — CHANGES_REQUESTED** [review]

File: Framework/DataObjects/inc/MantidDataObjects/MDGridBox.tcc
Comment (KedoKudo) [reply Framework/DataObjects/inc/MantidDataObjects/MDGridBox.tcc]: Converting `verticesContained` and `boxMightTouch` to vector has zero impact, but changing `distmaxs` and `distmins` to vector leads to error for all MDGridBox related tests (see the test log associated with 067256c below).

File: Framework/DataObjects/inc/MantidDataObjects/MDGridBox.tcc
Comment (martyngigg) [inline Framework/DataObjects/inc/MantidDataObjects/MDGridBox.tcc:1174]: This is very strange. The vector should act exactly as the array does - it feels like there is something about this code we don't fully understand. I don't want to hold up the bug fix but I think this should be investigated at some point and I think that would have to be done by stepping through with a debugger to see what operates differently in both scenarios.

Comment (martyngigg) [reply Framework/DataObjects/inc/MantidDataObjects/MDGridBox.tcc:1174]: Could we remove the commented out code for now to avoid confusion for future developers?

Comment (rosswhitfield) [reply Framework/DataObjects/inc/MantidDataObjects/MDGridBox.tcc:1174]: @KedoKudo `memset(distmins, 65535, numBoxes * sizeof(coord_t));` initializes the array as `-nan` and then every comparison with `distmins` will be `false`. I don't think this was what you were going for.

You can make using the vectors pass the unit tests If instead of

```cpp
std::vector<coord_t> distmins(numBoxes, radiusSquared + 1);
std::vector<coord_t> distmaxs(numBoxes, 0.0);
```
you do
```cpp
std::vector<coord_t> distmins(numBoxes, 0.0);
std::vector<coord_t> distmaxs(numBoxes, 0.0);
```
which suggest there is something wrong with the logic using `distmins`.

**Comment by rosswhitfield** [comment]
This new new approach looks much better and is given the correct results. I want to take another look tomorrow before I approve it.

File: Framework/DataObjects/inc/MantidDataObjects/MDGridBox.tcc
Comment (martyngigg) [reply Framework/DataObjects/inc/MantidDataObjects/MDGridBox.tcc]: Thanks @rosswhitfield. I didn't spot the bad initialization there - another reason to avoid C constructs!

**Review by rosswhitfield — APPROVED** [review]

**Review by martyngigg — APPROVED** [review]

**Pull request merged**
